### PR TITLE
Adding support for DBR 13.x and 14.x 

### LINKED ISSUE #1584: Unable to create `SparkLinker` on Databricks
### What happens?

The `SparkLinker` constructor fails on (AWS) Databricks with error: `py4j.Py4JException: Constructor com.databricks.libraries.JavaJarId([class java.net.URI, class java.lang.String, class java.lang.String]) does not exist`

### To Reproduce

Follow the "end-to-end" example in the `auto-data-linkage` [workshop](https://github.com/databricks-industry-solutions/auto-data-linkage/blob/f425125c62904f18f2e2f084467293f1878f908c/Workshop%20Materials/notebooks/Splink%20in%20Databricks%20-%20end-to-end%20example.py).

Specifically, connect to personal compute running DBR 13.3 LTS (with default libraries, init script, spark configuration, etc.) and run the following in a notebook:

```
%pip install splink --quiet

from splink.spark.spark_linker import SparkLinker
# or from splink.spark.linker import SparkLinker
from splink.databricks.enable_splink import enable_splink
import splink.spark.spark_comparison_library as cl

df = spark.read.table('main.default.<table>')

linker = SparkLinker(df, spark=spark) # error
```

### OS:

Databricks Runtime 13.3 LTS (Spark 3.4.1, Scala 2.12)

### Splink version:

3.9.6

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by boobay**
### Type of PR

- [X] BUG
- [ ] FEAT
- [ ] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
closes #1584

### Give a brief description for the solution you have provided
Adding new constructor arguments introduced in DBR 13.x and 14.x. 

13.x adds 2 new "optional" arguments
 - Option<String> apiToken
 - Option<String> workspaceApiUrl

14.x adds 1 new "optional" argument
 - Option<String> presignedUrl

### PR Checklist

- [x] Added documentation for changes
- [x] Added feature to example notebooks or tutorial (if appropriate)
- [x] Added tests (if appropriate)
- [x] Updated CHANGELOG.md (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

**Comment by RobinL** [comment]
Thanks for this.  I don't have access to Databricks, but I understand there have been recent changes that make it hard (or prevent?) registering arbitrary jars.  Does this code simply prevent  errors, or does it enable the custom Splink functions to work in later versions of databricks?

e.g. if you run `spark.sql("select jaro_winkler('john', 'jon')")` does it recognise the jaro_winkler function?

Either way I think it's useful, would just be good to understand what it does!

**Comment by boobay** [comment]
tldr; functions will work in later versions of Databricks :)

No worries. I understand. :)

I'll be honest. I am still learning to use the library for a current project, so bear with me. Just feel the need to be helpful, since I'm using your project and have the ability :)

When I manually run enable_splink, it successfully loads the jars (see log below)

`24/02/15 23:03:50 INFO SharedDriverContext: [Thread 391] attachLibrariesToSpark com.databricks.libraries.JavaJarId@99d74047
24/02/15 23:03:50 INFO SharedDriverContext: Attaching lib: file:/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/splink/files/spark_jars/scala-udf-similarity-0.1.1_spark3.x.jar to Spark
24/02/15 23:03:50 INFO LibraryDownloadManager: Downloading a library that was not in the cache: com.databricks.libraries.JavaJarId@99d74047
24/02/15 23:03:50 INFO LibraryDownloadManager: Attempt 1: wait until library com.databricks.libraries.JavaJarId@99d74047 is downloaded
24/02/15 23:03:50 INFO LibraryDownloadManager: Preparing to download library file using the default access mechanism: file:/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/splink/files/spark_jars/scala-udf-similarity-0.1.1_spark3.x.jar
24/02/15 23:03:50 INFO LibraryDownloadManager: Downloaded library com.databricks.libraries.JavaJarId@99d74047 as local file /local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar in 6 milliseconds
24/02/15 23:03:50 INFO SharedDriverContext: Successfully saved library com.databricks.libraries.JavaJarId@99d74047 to local file /local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar
24/02/15 23:03:50 INFO SparkContext: Added file /local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar at spark://10.77.86.34:40117/files/scala_udf_similarity_0_1_1_spark3_x.jar with timestamp 1708038230734
24/02/15 23:03:50 INFO Utils: Copying /local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar to /local_disk0/spark-81f2c6c3-2b53-4f0c-8f5a-a28726d59464/userFiles-7feae11e-5ecc-48b4-9814-83cb20bca95f/scala_udf_similarity_0_1_1_spark3_x.jar
24/02/15 23:03:50 INFO SparkContext: Added JAR /local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar at (spark://10.77.86.34:40117/jars/scala_udf_similarity_0_1_1_spark3_x.jar,Some(/local_disk0/tmp/addedFile5bca23870f7244e4992f8c8a6284bf6c2458553871709108914/scala_udf_similarity_0_1_1_spark3_x.jar)) with timestamp 1708038230749
24/02/15 23:03:50 INFO SharedDriverContext: Successfully attached library file:/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/splink/files/spark_jars/scala-udf-similarity-0.1.1_spark3.x.jar to Spark
24/02/15 23:03:50 INFO LibraryState: [Thread 391] Successfully attached library file:/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/splink/files/spark_jars/scala-udf-similarity-0.1.1_spark3.x.jar
`
I know the enable_splink functions ends up getting called in the SparkLinker (I think?) and subsequently registers the udfs a bit further down in the constructor.

**Comment by boobay** [comment]
Just for reference, the JAR common--libraries--libraries-spark_3.5_2.12_deploy.jar in /databricks/jars contains the JavaJarId class. I de-obfuscated it with IntelliJ and made the necessary changes.

You can usually find the appropriate file by grepping the artifacts for the classname.

**Comment by boobay** [comment]
The code (though pretty dirty with the try-except) should support runtimes up to 14.3. It'll try 14.3 constructor, then 13.3 and uses the original code if those both fail :)

**Comment by boobay** [comment]
@RobinL Out of curiosity, would you like any more information for the PR? I'm more than happy to provide information. Also, I'm just curious of a good way to get in-touch outside PR/issues? I'd like to ask some random questions about scale :D

**Comment by boobay** [comment]
Have a great leave. I sent you an email from my work address.

**Comment by RobinL** [comment]
Thank you - some researchers I was in contact with all tried this out in their databricks and it worked for them too 🎉

**Comment by Priebe1** [comment]
Hi @RobinL, do you know when this bugfix will be part of a new release? We are also using splink on databricks and would like to upgrade to a 14.x cluster
